crates/wasmtime/src/runtime/vm/gc/enabled/structref.rs

```diff
@@ -0,0 +1,471 @@
+        core::ptr::copy_nonoverlapping(le_bytes.as_ptr(), into, le_bytes.len());
+    }
+}
+unsafe impl PodValType for i8 {
+    fn read_le(le_bytes: &[u8]) -> Self {
+        i8::from_le_bytes(le_bytes.try_into().unwrap())
+    }
+    unsafe fn write_le(&self, into: *mut u8) {
+        let le_bytes = self.to_le_bytes();
+        core::ptr::copy_nonoverlapping(le_bytes.as_ptr(), into, le_bytes.len());
+    }
+}
+unsafe impl PodValType for i16 {
+    fn read_le(le_bytes: &[u8]) -> Self {
+        i16::from_le_bytes(le_bytes.try_into().unwrap())
+    }
+    unsafe fn write_le(&self, into: *mut u8) {
+        let le_bytes = self.to_le_bytes();
+        core::ptr::copy_nonoverlapping(le_bytes.as_ptr(), into, le_bytes.len());
+    }
+}
+unsafe impl PodValType for i32 {
+    fn read_le(le_bytes: &[u8]) -> Self {
+        i32::from_le_bytes(le_bytes.try_into().unwrap())
+    }
+    unsafe fn write_le(&self, into: *mut u8) {
+        let le_bytes = self.to_le_bytes();
+        core::ptr::copy_nonoverlapping(le_bytes.as_ptr(), into, le_bytes.len());
+    }
+}
+unsafe impl PodValType for i64 {
+    fn read_le(le_bytes: &[u8]) -> Self {
+        i64::from_le_bytes(le_bytes.try_into().unwrap())
+    }
+    unsafe fn write_le(&self, into: *mut u8) {
+        let le_bytes = self.to_le_bytes();
+        core::ptr::copy_nonoverlapping(le_bytes.as_ptr(), into, le_bytes.len());
+    }
+}
+unsafe impl PodValType for isize {
+    fn read_le(le_bytes: &[u8]) -> Self {
+        isize::from_le_bytes(le_bytes.try_into().unwrap())
+    }
+    unsafe fn write_le(&self, into: *mut u8) {
+        let le_bytes = self.to_le_bytes();
+        core::ptr::copy_nonoverlapping(le_bytes.as_ptr(), into, le_bytes.len());
+    }
+}
+
+unsafe impl PodValType for V128 {
+    fn read_le(le_bytes: &[u8]) -> Self {
+        let u128 = u128::from_le_bytes(le_bytes.try_into().unwrap());
+        u128.into()
+    }
+    unsafe fn write_le(&self, into: *mut u8) {
+        let le_bytes = self.as_u128().to_le_bytes();
+        core::ptr::copy_nonoverlapping(le_bytes.as_ptr(), into, le_bytes.len());
+    }
+}
+
+/// The backing storage for a GC-managed struct.
+///
+/// Methods on this type do not, generally, check against things like type
+/// mismatches or that the given offset to read from even falls on a field
+/// boundary. Omitting these checks is memory safe, due to our untrusted,
+/// indexed GC heaps. Providing incorrect offsets will result in general
+/// incorrectness, such as wrong answers or even panics, however.
+///
+/// Finally, these methods *will* panic on out-of-bounds accesses, either out of
+/// the GC heap's bounds or out of this struct's bounds. The former is necessary
+/// for preserving the memory safety of indexed GC heaps in the face of (for
+/// example) collector bugs, but the latter is just a defensive technique to
+/// catch bugs early and prevent action at a distance as much as possible.
+pub struct VMStructDataMut<'a> {
+    data: &'a mut [u8],
```

Comment:
To confirm, all data on the gc heap is always initialized, right? We don't have to use `MaybeUninit<u8>` here?

Reply:
From the Rust perspective: yes. We get it zeroed from mmap. At worst it has stale data from a since-reclaimed GC allocation.

From the Wasm perspective, it could be uninitialized, and we have to initialize it before we expose the object to Wasm.

But to answer the question of `MaybeUninit`: no, it shouldn't be necessary.